Code/Tools/AssetProcessor/native/unittests/AssetProcessingStateDataUnitTests.cpp

```diff
@@ -89,144 +113,120 @@ void AssetProcessingStateDataUnitTest::DataTest(AssetProcessor::AssetDatabaseCon
-        return;
-    }
+            return false;
+        };
 
-    //get all scan folders, there should only the two we added
-    scanFolders.clear();
-    UNIT_TEST_EXPECT_TRUE(stateData->GetScanFolders(scanFolders));
-    UNIT_TEST_EXPECT_TRUE(scanFolders.size() == 2);
-    UNIT_TEST_EXPECT_TRUE(ScanFoldersContainScanPath(scanFolders, "c:/O3DE/dev2"));
-    UNIT_TEST_EXPECT_TRUE(ScanFoldersContainScanPath(scanFolders, "c:/O3DE/game"));
-    UNIT_TEST_EXPECT_TRUE(ScanFoldersContainScanFolderID(scanFolders, scanFolder.m_scanFolderID));
-    UNIT_TEST_EXPECT_TRUE(ScanFoldersContainScanFolderID(scanFolders, gameScanFolderEntry.m_scanFolderID));
-
-    //remove scan folder by using a container
-    ScanFolderDatabaseEntryContainer tempScanFolderDatabaseEntryContainer; // note that on clang, its illegal to call a non-const function with a temp variable container as the param
-    UNIT_TEST_EXPECT_TRUE(stateData->RemoveScanFolders(tempScanFolderDatabaseEntryContainer)); // call with empty
-    UNIT_TEST_EXPECT_TRUE(stateData->RemoveScanFolders(scanFolders));
-    scanFolders.clear();
-    UNIT_TEST_EXPECT_FALSE(stateData->GetScanFolders(scanFolders));
-
-    ///////////////////////////////////////////////////////////
-    //setup for sources tests
-    //for the rest of the test lets add the original scan folder
-    scanFolder = ScanFolderDatabaseEntry("c:/O3DE/dev", "dev", "devkey2");
-    UNIT_TEST_EXPECT_TRUE(stateData->SetScanFolder(scanFolder));
-    ///////////////////////////////////////////////////////////
-
-    //////////////////////////////////////////////////////////////////////////
-    //Sources
-    auto SourcesContainSourceID = [](const SourceDatabaseEntryContainer& sources, AZ::s64 sourceID) -> bool
+        //there are no scan folders yet so trying to find one should fail
+        EXPECT_FALSE(m_connection.GetScanFolders(scanFolders));
+        EXPECT_FALSE(m_connection.GetScanFolderByScanFolderID(0, scanFolder));
+        EXPECT_FALSE(m_connection.GetScanFolderBySourceID(0, scanFolder));
+        EXPECT_FALSE(m_connection.GetScanFolderByProductID(0, scanFolder));
+        EXPECT_FALSE(m_connection.GetScanFolderByPortableKey("sadfsadfsadfsadfs", scanFolder));
+        scanFolders.clear();
+
+        //add a scanfolder
+        scanFolder = ScanFolderDatabaseEntry("c:/O3DE/dev", "dev", "rootportkey");
+        EXPECT_TRUE(m_connection.SetScanFolder(scanFolder));
+        EXPECT_NE(scanFolder.m_scanFolderID, AzToolsFramework::AssetDatabase::InvalidEntryId);
+
+        //add the same folder again, should not add another because it already exists, so we should get the same id
+        // not only that, but the path should update.
+        ScanFolderDatabaseEntry dupeScanFolder("c:/O3DE/dev2", "dev", "rootportkey");
+        dupeScanFolder.m_scanFolderID = AzToolsFramework::AssetDatabase::InvalidEntryId;
+        EXPECT_TRUE(m_connection.SetScanFolder(dupeScanFolder));
+        EXPECT_TRUE(dupeScanFolder == scanFolder);
```

Comment:
could do this instead?
`EXPECT_TRUE(dupeScanFolder == scanFolder) << "AssetProcessingStateDataTest Failed - scan folder failed to add";`

Reply:
Yeah, that seems better. I didn't know that we can throw error messages in this way. I'll make the change.